projects/lights_control/src/lights_can.c

```diff
@@ -0,0 +1,89 @@
+#include "can.h"
+#include "can_hw.h"
+#include "can_msg_defs.h"
+#include "can_unpack.h"
+#include "event_queue.h"
+#include "status.h"
+
+#include "exported_enums.h"
+#include "lights_can.h"
+#include "lights_events.h"
+
+// RX handler for signal type events.
+static StatusCode prv_rx_signal_handler(const CANMessage *msg, void *context,
+                                        CANAckStatus *ack_reply) {
+  uint8_t signal_type = 0;
+  uint8_t state = 0;
+  status_ok_or_return(CAN_UNPACK_LIGHTS_SIGNAL_TYPES(msg, &signal_type, &state));
+  if (signal_type >= NUM_LIGHTS_SIGNAL_TYPES) {
+    return status_msg(STATUS_CODE_INVALID_ARGS, "Unsupported signals type.");
+  }
+  return event_raise((state) ? LIGHTS_EVENT_SIGNAL_ON : LIGHTS_EVENT_SIGNAL_OFF,
+                     (LightsEventSignalMode)signal_type);
+}
+
+// RX handler for generic lights.
+static StatusCode prv_rx_generic_lights_handler(const CANMessage *msg, void *context,
+                                                CANAckStatus *ack_reply) {
+  LightsCanSettings *settings = (LightsCanSettings *)context;
+  uint8_t light_type = 0;
+  uint8_t state = 0;
+  status_ok_or_return(CAN_UNPACK_LIGHTS_GENERIC_TYPES(msg, &light_type, &state));
+  if (light_type >= NUM_LIGHTS_GENERIC_TYPES) {
+    return status_msg(STATUS_CODE_INVALID_ARGS, "Unsupported generic light type.");
+  }
+  return event_raise((state) ? LIGHTS_EVENT_GPIO_ON : LIGHTS_EVENT_GPIO_OFF,
+                     (LightsEventSignalMode)settings->generic_light_lookup[light_type]);
+}
+
+// RX handler for sync.
+static StatusCode prv_rx_sync_handler(const CANMessage *msg, void *context,
+                                      CANAckStatus *ack_reply) {
+  status_ok_or_return(CAN_UNPACK_LIGHTS_SYNC(msg));
+  return event_raise(LIGHTS_EVENT_SYNC, 0);
+}
+
+// RX handler for BPS faults.
+static StatusCode prv_rx_bps_fault_handler(const CANMessage *msg, void *context,
+                                           CANAckStatus *ack_reply) {
+  uint8_t status = 0;
+  status_ok_or_return(CAN_UNPACK_BPS_HEARTBEAT(msg, &status));
+  return event_raise(LIGHTS_EVENT_BPS, status);
+}
+
+// RX handler for horn.
+static StatusCode prv_rx_horn_handler(const CANMessage *msg, void *context,
+                                      CANAckStatus *ack_reply) {
+  uint8_t state = 0;
+  status_ok_or_return(CAN_UNPACK_HORN(msg, &state));
+  return event_raise((state) ? LIGHTS_EVENT_GPIO_ON : LIGHTS_EVENT_GPIO_OFF,
+                     LIGHTS_EVENT_GPIO_PERIPHERAL_HORN);
+}
+
+StatusCode lights_can_init(const LightsCanSettings *settings, LightsCanStorage *storage) {
+  CANSettings can_settings = {
+    .device_id = settings->device_id,
+    .bitrate = CAN_HW_BITRATE_125KBPS,
```

Comment:
Why are you hardcoding your bitrate? We haven't decided on a system network speed yet.

Reply:
I'll pass it as config.